### PR TITLE
Add HelpopMessageSentEvent

### LINKED ISSUE #5479: An event fired when a new message is received through /helpop command
### Feature description

I suggest to fire an event when messages are sent through the /helpop command.

### How the feature is useful

Such an event would allow developers to process help messages and redirect them where they want.

### PR DESCRIPTION
**Pull request by YanisBft**
### Information

This PR closes #5479. 

### Details

**Proposed feature:**  

This pull request adds an event fired when a new message has been sent to the helpop channel.

**Environments tested:**    

OS: Ubuntu 22.04

Java version:  OpenJDK 18

- [X] Most recent Paper version (1.20.1, git-Paper-130)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**

### CODE REVIEW

File: Essentials/src/main/java/net/ess3/api/events/HelpopMessageSentEvent.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package net.ess3.api.events;
```
Comment (JRoy) [inline Essentials/src/main/java/net/ess3/api/events/HelpopMessageSentEvent.java]: new events should be added to the `net.essentialsx.api.v2.events` package.

File: Essentials/src/main/java/net/ess3/api/events/HelpopMessageSentEvent.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public String getMessage() {
+        return message;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
```
Comment (JRoy) [inline Essentials/src/main/java/net/ess3/api/events/HelpopMessageSentEvent.java]:
```suggestion
    public IMessageRecipient getSender() {
        return sender;
    }

    public String getMessage() {
        return message;
    }

    @Override
    public HandlerList getHandlers() {
        return handlers;
    }
    
    public static HandlerList getHandlerList() {
        return handlers;
    }
```

File: Essentials/src/main/java/net/ess3/api/events/HelpopMessageSentEvent.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.earth2me.essentials.messaging.IMessageRecipient;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+/**
+ * Called after a message has been sent to the helpop channel.
+ */
+public class HelpopMessageSentEvent extends Event {
+
```
Comment (JRoy) [inline Essentials/src/main/java/net/ess3/api/events/HelpopMessageSentEvent.java]:
```suggestion
```

**Review by JRoy — CHANGES_REQUESTED** [review]
few things, also i would also like a modifiable list of recipients for who receives the helpop message.

**Comment by YanisBft** [comment]
> few things, also i would also like a modifiable list of recipients for who receives the helpop message.

Done it in 066d6b1, let me know if it was not the right way to do it :)

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java [hunk 4]
```diff
@@ -20,25 +23,37 @@ public Commandhelpop() {
...
+        ess.getServer().getPluginManager().callEvent(sentEvent);
+
+        return finalMessage;
+    }
+
+    private List<? extends IMessageRecipient> getOnlineHelpopRecipients() {
+        return ess.getOnlinePlayers().stream()
+                .filter(player -> player.hasPermission("essentials.helpop.receive"))
+                .map(ess::getUser)
+                .collect(Collectors.toList());
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java:56]: don't use streams and this list isn't mutable. changes to the list should effect who is sent the message.

Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: don't use streams and this list isn't mutable. changes to the list should effect who is sent the message.

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java (hunk 4 above)
Comment (YanisBft) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: This event is currently called after the message has been sent, and I think you want to have an event before it is sent. What should I do?

Comment (JRoy) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: Correct, make it fire before. I guess you could rename the event to `HelpopMessageSendEvent` to better illustrate that. Unless there is a benefit to have the event send after that I don't see?

Comment (YanisBft) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: If it's not cancellable I don't see any difference either. Will do it soon

Comment (YanisBft) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: Still, how should I make a modifiable list of recipients properly?

Comment (JRoy) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: > Still, how should I make a modifiable list of recipients properly?

create an array list, do a for loop with all online players, add any player with the permission to the list, use the list in the event, call the event, then use the list to send the message out.

Comment (YanisBft) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: Done!

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java [hunk 5]
```diff
@@ -20,25 +23,39 @@ public Commandhelpop() {
...
+        for (User user : ess.getOnlineUsers()) {
+            if (user.getBase().hasPermission("essentials.helpop.receive")) {
+                recipients.add(user);
+            }
+        }
+
+        final HelpopMessageSendEvent sendEvent = new HelpopMessageSendEvent(from, recipients, message);
+        ess.getServer().getPluginManager().callEvent(sendEvent);
+
+        sendEvent.getRecipients().forEach(recipient -> recipient.sendMessage(finalMessage));
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: just use a regular for loop

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java [hunk 6]
```diff
@@ -20,25 +23,39 @@ public Commandhelpop() {
...
-        final String message = tl("helpOp", from, FormatUtil.stripFormat(getFinalArg(args, 0)));
-        ess.getLogger().log(Level.INFO, message);
-        ess.broadcastMessage("essentials.helpop.receive", message);
-        return message;
+
+        final String message = FormatUtil.stripFormat(getFinalArg(args, 0));
+        final String finalMessage = tl("helpOp", from.getDisplayName(), message);
+        ess.getLogger().log(Level.INFO, finalMessage);
+
+        final List<IMessageRecipient> recipients = new ArrayList<>();
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandhelpop.java]: would prefer the type for this list be `net.ess3.api.IUser`

File: Essentials/src/main/java/net/essentialsx/api/v2/events/HelpopMessageSendEvent.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    public List<IMessageRecipient> getRecipients() {
+        return recipients;
+    }
+
+    /**
+     * Sets the recipients of the helpop message.
+     * @param recipients the recipients.
+     */
+    public void setRecipients(final List<IMessageRecipient> recipients) {
```
Comment (JRoy) [inline Essentials/src/main/java/net/essentialsx/api/v2/events/HelpopMessageSendEvent.java]: don't need a set method, people can just modify the list

File: Essentials/src/main/java/net/essentialsx/api/v2/events/HelpopMessageSendEvent.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.List;
+
+/**
+ * Called just before a message is sent to the helpop channel.
+ */
+public class HelpopMessageSendEvent extends Event {
+    private static final HandlerList handlers = new HandlerList();
+
+    private final IMessageRecipient sender;
+    private List<IMessageRecipient> recipients;
```
Comment (JRoy) [inline Essentials/src/main/java/net/essentialsx/api/v2/events/HelpopMessageSendEvent.java]: this can be final

**Review by JRoy — CHANGES_REQUESTED** [review]

**Review by JRoy — APPROVED** [review]

**Pull request merged**